sled-agent/src/bootstrap/early_networking.rs

```diff
@@ -38,59 +47,238 @@ pub enum EarlyNetworkSetupError {
     Dendrite(String),
 
     #[error("Error during DNS lookup: {0}")]
-    DnsResolver(#[from] internal_dns::resolver::ResolveError),
+    DnsResolver(#[from] ResolveError),
+}
+
+enum LookupSwitchZoneAddrsResult {
+    // We found every switch zone reported by internal DNS.
+    TotalSuccess(HashMap<SwitchLocation, Ipv6Addr>),
+    // We found some (but not all) switch zones reported by internal DNS.
+    PartialSuccess(HashMap<SwitchLocation, Ipv6Addr>),
 }
 
 /// Code for configuring the necessary network bits to bring up the control
 /// plane
-pub struct EarlyNetworkSetup {
-    log: Logger,
+pub struct EarlyNetworkSetup<'a> {
+    log: &'a Logger,
 }
 
-impl EarlyNetworkSetup {
-    pub fn new(log: &Logger) -> Self {
-        EarlyNetworkSetup { log: log.clone() }
+impl<'a> EarlyNetworkSetup<'a> {
+    pub fn new(log: &'a Logger) -> Self {
+        EarlyNetworkSetup { log }
+    }
+
+    pub async fn lookup_boundary_switch_addrs(
+        &self,
+        resolver: &DnsResolver,
+        config: &RackNetworkConfig,
+    ) -> HashSet<Ipv6Addr> {
+        let switch_zone_addrs = self.lookup_switch_zone_addrs(resolver).await;
+        let mut boundary_switch_addrs = HashSet::new();
+
+        for uplink_config in &config.uplinks {
+            if let Some(&zone_addr) =
+                switch_zone_addrs.get(&uplink_config.switch)
+            {
+                boundary_switch_addrs.insert(zone_addr);
+            } else {
+                // `lookup_boundary_switch_addrs` is best effort: if one of the
+                // switches is offline (or we can't find it for any number of
+                // reasons), we may not have the address for this
+                // `uplink_config`. Just log a warning and move on.
+                warn!(
+                    self.log,
+                    "No switch zone address found for {:?}",
+                    uplink_config.switch,
+                );
+            }
+        }
+
+        // TODO-correctness: Could `boundary_switch_addrs` be _empty_ here?
+        // `lookup_switch_zone_addrs` does not return unless it finds at least
+        // one switch location, so the only way this could happen is if there
+        // are no uplinks configured at all, or there are only uplinks
+        // configured for one switch location and we found the other one. Both
+        // of those are misconfiguration issues that we cannot overcome here, so
+        // we will return what we have.
+        boundary_switch_addrs
+    }
+
+    pub async fn lookup_switch_zone_addrs(
```

Comment:
It may be clarifying to use `underlay_addrs` for these function name suffixes, rather than just `_addrs`. I realize this is mostly copied, but for the uninitiated it may not be clear that we are not looking up `bootstrap` addresses given this is in `EarlyNetworking`.